# Arborele de decizie

### Importăm librăriile necesare

In [2]:
# Încărcarea librăriilor
import pandas as pd # Dataframe-ul și series-ul nostru
from sklearn.tree import DecisionTreeClassifier # Importăm Clasificatorul „Arborele decizional”
from sklearn.model_selection import train_test_split # Importăm funcția de separare a datelor în date de antrenare și date de testare
from sklearn import metrics #Importăm modulel metrici pentru caluclarea acurateții

### Importăm setul de date


Eu voi lucra cu data set-ul Breast Cancer
Labels:
1=Healthy controls
2=Patients

In [3]:
# încărcarea setului de date
BC = pd.read_csv("data/dataR2.csv")
BC.iloc[44]

Age                71.000000
BMI                30.300000
Glucose           102.000000
Insulin             8.340000
HOMA                2.098344
Leptin             56.502000
Adiponectin         8.130000
Resistin            4.298900
MCP.1             200.976000
Classification      1.000000
Name: 44, dtype: float64

In [4]:
#despărțim datasetul în vector de caracteristici (features) și vectorul clasă
features = [ 'Age','BMI','Glucose','Insulin','HOMA','Leptin','Adiponectin','Resistin','MCP.1']
X = BC[features] # caracteristicile (atributele pentru X)
y = BC.Classification # Variabila clasă (independentă dacă facem analogie cu regresia)

### Împărțirea setului de date
Pentru a evalua performanța modelului o strategie ar fie împărțirea setului de date în: date de antrenare/învățare și date pentru testare/evaluare.

Putem face acest lucru utilizând funcția train_test_split(). Funcția ia 3 parametri: caracteristicile, clasa și lungimea setului de date pentru testare.

In [5]:
# Împărțim setul de date în train set și test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) # 70% pentru antrenare and 30% testare
X_test

Age        BMI  Glucose  Insulin       HOMA   Leptin  Adiponectin  \
95    49  29.777778       70    8.396   1.449709  51.3387    10.731740   
44    71  30.300000      102    8.340   2.098344  56.5020     8.130000   
56    42  21.359915       93    2.999   0.687971  19.0826     8.462915   
97    40  27.636054      103    2.432   0.617890  14.3224     6.783870   
69    44  19.560000      114   15.890   4.468268  13.0800    20.370000   
114   72  25.590000       82    2.820   0.570392  24.9600    33.750000   
73    72  23.620000      105    4.420   1.144780  21.7800    17.860000   
58    51  19.132653       93    4.364   1.001102  11.0816     5.807620   
35    67  29.606767       79    5.819   1.133929  21.9033     2.194280   
38    60  26.349292      103    5.138   1.305395  24.2998     2.194280   
65    45  23.140496      116    4.902   1.402626  17.9973     4.294705   
96    44  27.887617       99    9.208   2.248594  12.6757     5.478170   
2     82  23.124670       91    4.498   1.009651  17.9393    22.432040   
78    86  26.666667      201   41.611  20.630734  47.6470     5.357135   
109   75  30.480000      152    7.010   2.628283  50.5300    10.060000   
53    45  20.829995       74    4.560   0.832352   7.7529     8.237405   
102   65  30.915577       97   10.491   2.510147  44.0217     3.710090   
99    69  28.444444      108    8.808   2.346451  14.7485     5.288025   
94    60  31.231410      131   30.130   9.736007  37.8430     8.404430   
42    75  27.300000       85    5.197   1.089638  10.3900     9.000805   
104   57  34.838148       95   12.548   2.940415  33.1612     2.364950   
17    61  32.038959       85   18.077   3.790144  30.7729     7.780255   
80    59  28.672626       77    3.188   0.605507  17.0220    16.440480   
54    49  20.956608       94   12.305   2.853119  11.2406     8.412175   
93    49  32.461911      134   24.887   8.225983  42.3914    10.793940   
33    43  34.422174       89   23.194   5.091856  31.2128     8.300955   
67    64  22.222222       98    5.700   1.377880  12.1905     4.783985   
48    69  29.400000       89   10.704   2.349885  45.2720     8.286300   
46    75  25.700000       94    8.079   1.873251  65.9260     3.741220   
92    52  30.801249       87   30.212   6.483495  29.2739     6.268540   
31    53  36.790166      101   10.175   2.534932  27.1841    20.030000   
55    34  24.242424       92   21.699   4.924226  16.7353    21.823745   
59    62  22.656250       92    3.482   0.790182   9.8648    11.236235   
10    34  21.470000       78    3.469   0.667436  14.5700    13.110000   
32    28  35.855815       87    8.576   1.840410  68.5102     4.794200   

     Resistin     MCP.1  
95   20.76801   602.486  
44    4.29890   200.976  
56   17.37615   321.919  
97   26.01360   293.123  
69    4.62000   220.660  
114   3.27000   392.460  
73    4.82000   195.940  
58    5.57055    90.600  
35    4.20750   585.307  
38   20.25350   378.996  
65    5.26330   518.586  
96   23.03306   407.206  
2     9.27715   554.697  
78   24.37010  1698.440  
109  11.73000    99.450  
53   28.03230   382.955  
102  20.46850   396.648  
99   16.48508   353.568  
94   11.50005   396.021  
42    7.57670   335.393  
104   9.95420   655.834  
17   13.68392   444.395  
80   31.69040   910.489  
54   23.11770   573.630  
93    5.76800   656.393  
33    6.71026   960.246  
67   13.91245   395.976  
48    4.53000   215.769  
46    4.49685   206.802  
92   24.24591   764.667  
31   10.26309   695.754  
55   12.06534   481.949  
59   10.69548   703.973  
10    6.92000   354.600  
32   21.44366   358.624

### Construirea modelului

In [6]:
# Crearea obiectului de clasificare pe bază de arbore de decizie cu parametru criterie
clasificator = DecisionTreeClassifier(criterion="entropy")

# Antrenarea clasificatorului cu datele noastre utilizând metoda fit() și ca parametru avem datele de antrenare
clasificator_arbore_decizie = clasificator.fit(X_train,y_train)

### Evaluarea modelului

In [7]:
#Prezice clasele pentru setul de testare utilizând metoda predict(parametru_setu_de_testare)
y_prezis = clasificator_arbore_decizie.predict(X_test)


# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_prezis))
print(y_test, y_prezis)

Accuracy: 0.7714285714285715
95     2
44     1
56     2
97     2
69     2
114    2
73     2
58     2
35     1
38     1
65     2
96     2
2      1
78     2
109    2
53     2
102    2
99     2
94     2
42     1
104    2
17     1
80     2
54     2
93     2
33     1
67     2
48     1
46     1
92     2
31     1
55     2
59     2
10     1
32     1
Name: Classification, dtype: int64 [2 2 2 2 2 1 2 2 1 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 1 1 1 2 1 1 1 1]


In [8]:
obiect_nou = [[45,29.38475666,90,4.713,1.046286,23.8479,6.644245,15.55625,621.273]]

raspuns=clasificator_arbore_decizie.predict(obiect_nou)

print(raspuns)

[2]


### Vizualizarea arborelui

Puteți folosi funcția export_graphviz din Scikit-learn pentru a afișa arborele în Jupyter Notebook. Pentru a face diagram de asemenea este nevoie de biblioteca pudotplus.
Pentru instalarea clasică:

- pip install graphviz

- pip install pydotplus

funcția export_graphviz transformă clasificatorul nostru intr-un fișier cu puncte, iar pydotplus convertește aceste puncte într-un fișier png.

In [9]:
from sklearn.externals.six import StringIO  
from IPython.display import Image  
from sklearn.tree import export_graphviz
import pydotplus


c:\users\dp\envs\machinelearning\lib\site-packages\sklearn\externals\six.py:31: DeprecationWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", DeprecationWarning)


In [10]:
dot_data = StringIO()
export_graphviz(clasificator_arbore_decizie, out_file=dot_data,  
                filled=True, rounded=True,
                special_characters=True, feature_names = features,class_names=['0','1'])
graph = pydotplus.graph_from_dot_data(dot_data.getvalue())  


In [11]:
graph.write_png('weather.png')
Image(graph.create_png())

InvocationException: GraphViz's executables not found

Concluzii :
Pentru lucrarea de laborator numarul 4 am ales un dataset despre cancerul mamar, pe baza a mai multor criterii. Aceasta problema am realizat-o cu ajutorul arborilor de decizii.

Pentru a evalua performanța modelului am împărțit setul de date în: date de antrenare/învățare 70% și date pentru testare/evaluare 30%.

Putem spune ca arborele obtinut in final ne permite prezicerea cancerului mamar cu   Accuracuratete de : 0.7714285714285715 ceea ce este un procentaj destul de bun.
